### PR TITLE
feat: refactor sidebar layout

### PR DESCRIPTION
**Pull request by zchsh**
## "Ready for Review" checklist

- [x] The Vercel preview link has been added to this PR's description
- [x] The Asana task has been added to this PR's description
- [x] This PR's link has been added to the "📐 Pull Request" field on the Asana task
- [x] This Vercel preview link has been added to the "🔍 Preview" field on the Asana task
- [x] The description template has been filled in below

---

## Relevant links

- [Preview link](https://dev-portal-git-zsrefactor-sidebar-layout-hashicorp.vercel.app/waypoint) 🔎
- [Asana task](https://app.asana.com/0/1100423001970639/1201625161290271/f) 🎟️

## What

This PR refactors the `docs-layout` so that page content is located in the main document's scrolling area (rather than in a separate scroll container as in the current implementation).

## Why

NextJS's client-side routing does not play nicely with scrollable content not in the main document's scrolling area. For example, we do not see the content jump back to the top of the page on navigation as we'd expect.

## How

A new partial, `sidebar-sidecar-layout`, has been created within `docs-layout`. This partial handles layout positioning concerns only.

A demo of this partial can be seen at [/_layout-dev](https://dev-portal-git-zsrefactor-sidebar-layout-hashicorp.vercel.app/_layout-dev). (This demo will be removed before merging this PR; it felt useful for dev and review purposes).

## Testing

- [ ] Test out the simplified demo example of `sidebar-sidecar-layout` at [/_layout-dev](https://dev-portal-git-zsrefactor-sidebar-layout-hashicorp.vercel.app/_layout-dev), toggling the various content lengths in each of the `sidebar`, `main`, and `sidecar` areas. Ensure the behaviour is what you expect.
- [ ] Visit pages that use the layout, such as [`/waypoint`](https://dev-portal-git-zsrefactor-sidebar-layout-hashicorp.vercel.app/waypoint) and [`/waypoint/docs`](https://dev-portal-git-zsrefactor-sidebar-layout-hashicorp.vercel.app/waypoint/docs) pages. Ensure the layout is behaving as expected with real docs content.

## Anything else?

To reduce brittleness in `SidebarSidecarLayout`, we could observe the height of the sticky `header` in `BaseNewLayout`.  This would eliminate the need for global CSS variables that try to statically capture the height of potentially dynamic content. Have noted this caveat in code comments, and deferring this work until later iterations.

### CODE REVIEW

File: src/components/dev-alert-banner/index.tsx
Comment (zchsh) [inline src/components/dev-alert-banner/index.tsx:3]: Added `DevAlertBanner` to decouple the chained `docs-layout` >> `base-new-layout` while not repeating this code in both layouts.

File: src/components/sidebar/style.module.css
Comment (zchsh) [inline src/components/sidebar/style.module.css:2]: `box-shadow` and other sidebar layout styles are now set in `sidebar-sidecar-layout`.

File: src/components/sidecar/style.module.css
Comment (zchsh) [inline src/components/sidecar/style.module.css:8]: Positioning and width styles no longer needed, `sidebar-sidecar-layout` handles these, and we just need to drop in content.

File: src/layouts/base-new/index.tsx
Comment (zchsh) [inline src/layouts/base-new/index.tsx]: Abstracted `DevAlertBanner` so it could be re-used in `DocsLayout` without having to use this layout as the underlying layout.

File: src/layouts/base-new/base-new-layout.module.css
Comment (zchsh) [inline src/layouts/base-new/base-new-layout.module.css:4]: Have used `Min100Layout` as it implements this same functionality. Added `.header` class to ensure header bars are sticky in this layout.

File: src/layouts/docs/docs-layout.module.css
Comment (zchsh) [inline src/layouts/docs/docs-layout.module.css:51]: The content width and positioning are now handled by `sidebar-sidecar-layout`, this `main` content is responsible for content only (and for its semantic role, since `sidebar-sidecar-layout` just uses divs).

File: src/layouts/docs/partials/sidebar-sidecar-layout/index.tsx
Comment (zchsh) [inline src/layouts/docs/partials/sidebar-sidecar-layout/index.tsx:15]: We could, if we wanted to, roll this layout right into the `docs-layout` component, since it isn't exactly meant to be used on its own. I found it very helpful to split it out when developing though, so figure it might make sense to keep it split it for maintenance (especially while we're still figuring out docs-layout content styling).

File: src/layouts/docs/index.tsx
Comment (zchsh) [inline src/layouts/docs/index.tsx:34]: Have simplified `docs-layout` a bit by leveraging a new `sidebar-sidecar-layout` component, which handles the layout positioning concerns.

File: src/pages/_layout-dev/index.tsx
Comment (zchsh) [inline src/pages/_layout-dev/index.tsx:19]: Intention is to delete this, started this as a skateboard version of getting the layout working the way I think we want it to. Leaving it up for now to make it easier to isolate and debug layout arrangement concerns from docs page things.

File: src/pages/_layout-dev/style.module.css
Comment (zchsh) [inline src/pages/_layout-dev/style.module.css:1]: As mentioned in accompanying `.tsx` file, intention is to delete this once we're done reviewing and have the layout work approved. Though also considering trying to move this into Swingset or something (though not exactly sure how I'll do that, yet).

File: src/pages/index.module.css
Comment (zchsh) [inline src/pages/index.module.css:10]: Intent is to remove this as mentioned in the TODO, just wanted to demo the `BaseNewLayout` working with longer content (so sticky header behaviour could be confirmed)

File: src/pages/index.tsx
Comment (zchsh) [inline src/pages/index.tsx:5]: `Footer` now handled by `BaseNewLayout`. I think I might've handled this in the Vault `/docs` PR as well, but hoping to merge this PR first, I think, so will rebase the `/docs` work onto this.

File: src/pages/style.css
Comment (zchsh) [inline src/pages/style.css:13]: File is now empty, so baleeted.

File: src/styles/sizing.css
Comment (zchsh) [inline src/styles/sizing.css:4]: As this comment mentions, the one thing I think is objectively worse about the `body`-as-scrolling layout approach is that it requires we know the height of the sticky nav in order to implement the sidebar behaviour design has given us`*`. The simple-but-not-robust solution to that problem is to have hard-coded CSS vars as we have here.

This hard-coded approach kind of forces us to make our layouts brittle in ways that will affect users who change their default font-size. I think a more robust solution here would be to observe the sticky bar height and dynamically update the calculated height, to avoid that brittleness (as mentioned in the `sidebar-sidecar-layout/style.module.css`).

But, holding off on the not-hard-coded enhancement for now - want to get a higher-level go-ahead that this layout approach makes sense first! 🙇 

> `*` CAVEAT - seems required as far as I can tell - we had to run this kind of layout in Learn and ran into a very similar conundrum... in that case we've just stuck with magic static numbers for the nav bar heights, I think, but it's always kinda worried me.

File: src/components/sidebar/components/sidebar-nav/style.module.css
Comment (BRKalow) [inline src/components/sidebar/components/sidebar-nav/style.module.css:6]: 👀  seems the focus outline is getting clipped:
Screen Shot 2022-01-12 at 5 22 20 PM

**Comment by BRKalow** [comment]
I'd like to see if we could continue retaining the layout composition of `BaseNewLayout -> SidebarSidecarLayout`. Right now it looks like we've got duplicate styles across both, as well as duplicative implementation of `<NavigationHeader />` and `<DevAlertBanner />`. Curious to hear why you opted to decouple from the base layout!

**Comment by zchsh** [comment]
@BRKalow Happy to give the re-coupling a shot - opted to decouple from the base layout as a WIP step while working through the complexities of the `sidebar-sidecar-layout`.

In order to keep the `BaseNewLayout -> SidebarSidecarLayout` composition, I think `SidebarSidecarLayout` will need to be aware of the sticky header height (hopefully not through magic numbers, but maybe that's okay?), and we'll therefore need to have the height observer logic in `BaseNewLayout`. 

Do you have any thoughts on the height observer issue? Eg are the magic `--alert-bar-height` & `--navigation-header-height` feel acceptable, or does implementing a height observer for the sticky bar seem preferable?

**Comment by BRKalow** [comment]
I think in the interest of iteration it'd okay to use static numbers for now! We can introduce the height observer down the line if it feels like the right move.

**Comment by zchsh** [comment]
@BRKalow sounds good - will keep the global CSS vars then, and re-couple to `BaseNewLayout` 👍 (and make the footer optional in `BaseNewLayout` as it was before).

File: src/components/sidebar/components/sidebar-nav/style.module.css
Comment (zchsh) [reply src/components/sidebar/components/sidebar-nav/style.module.css:6]: Ah dang... when I don't have `overflow: hidden`, we end up with a massive scroll issue due to how `Skip to main content` is positioned by default.

@ashleemboyer Would an acceptable resolution path here to be position the `Skip to main content` at `top: 4; left: 4` (when focused) in order to avoid this cutoff?

File: src/pages/index.module.css
Comment (zchsh) [reply src/pages/index.module.css]: ✅ Removed

**Comment by zchsh** [comment]
@BRKalow Have refactored so that `SidebarSidecarLayout` leverages `BaseNewLayout`, and have left in the static CSS vars. @ashleemboyer tagged you in the `SkipToMainContent` positioning issue, hopefully that fix looks okay 🙇

**Review by BRKalow — APPROVED** [review]
This looks solid to me. Preview looks good, I tested the scroll behavior and all seems to be working like we would expect! Thanks for the detailed comments in the CSS!

**Review by ashleemboyer — CHANGES_REQUESTED** [review]
In addition to the individual comments I've left, I have some other concerns that were harder to pinpoint where to place comments. There were some margin, padding, and width changes made that don't match [the grids defined in Figma](https://www.figma.com/file/VD7ahvXuXWJApeGnhbW4hv/?node-id=1531%3A40889), such as some spacing that was built from margin + padding was combined into just margin and the max-width of the main content area was changed. I think it'll be easier to chat about these things either on Slack or Zoom, so let me know when you want to do that @zchsh!

File: src/components/sidebar/components/sidebar-nav/style.module.css
Comment (zchsh) [reply src/components/sidebar/components/sidebar-nav/style.module.css:6]: Absolutely - here's a screen recording showing the horizontal scroll issue when `overflow: hidden` is removed:

https://user-images.githubusercontent.com/4624598/151180142-d2f2c5e8-338c-4a8a-af2b-2f75a6a95677.mp4

Screen recording is in Safari, but I'm seeing the issue in Firefox and Chrome as well. The `top` & `left` set to `0` is minor compared to the overflow issue - was done as a fix for the cutoff Bryce mentioned above _after_ adding `overflow: hidden` to s

overflow-hidden-issue
olve for the horizontal scrolling issue:

File: src/layouts/docs/partials/sidebar-sidecar-layout/style.module.css
Comment (zchsh) [reply src/layouts/docs/partials/sidebar-sidecar-layout/style.module.css:25]: I think they do have something to do with the layout - the `sidebar` part of the layout creates a scroll container and positions itself to take up the expected height portion of the screen, even if its contents aren't that tall. The shadow etc need to be applied at that level, I think, rather than within the scroll container. Moving the `box-shadow` & `background-color` into `components/sidebar` would mean we'd need to break that model, ie rather than just having the sidebar contents, `components/sidebar` would then need to take on some of these other layout concern styles as well (`max-height`, `width`, etc).

Let me know if that layout vs content rationale seems reasonable 🙇‍♂️

File: src/layouts/docs/partials/sidebar-sidecar-layout/style.module.css
Comment (zchsh) [reply src/layouts/docs/partials/sidebar-sidecar-layout/style.module.css:33]: Great call - yep, will update! 👍

Comment (zchsh) [reply src/layouts/docs/partials/sidebar-sidecar-layout/style.module.css]: ✅ Updated

File: src/layouts/docs/index.tsx
Comment (zchsh) [reply src/layouts/docs/index.tsx]: Thanks for catching this - ✅ have moved adjacent to `main` 👍

File: src/layouts/docs/partials/sidebar-sidecar-layout/style.module.css
Comment (ashleemboyer) [reply src/layouts/docs/partials/sidebar-sidecar-layout/style.module.css:25]: I considered the `background-color` and `box-shadow` to be properties that are set regardless of the sidebar's positioning in the layout, things that are more "constant" across usage rather than dependent on the layout. Thinking some more on it `box-shadow`, would be different if in some future layout we put the sidebar on the right side of the page, so that probably is layout dependent! `background-color` seems like a constant though, and like something that should be delegated to the sidebar component to set separate from the layout.

File: src/components/sidebar/components/sidebar-nav/style.module.css
Comment (ashleemboyer) [reply src/components/sidebar/components/sidebar-nav/style.module.css:6]: Ah! I see now! So there are a couple of things we can do here instead of `overflow: hidden` in this class.
- Add `overflow-x: hidden` to the `.sidebar` class in `src/layouts/docs/partials/sidebar-sidecar-layout/style.module.css` (it used to be in the `.sidebar` class defined in `src/components/sidebar/style.module.css`).
- Change the default value of `right` in the `.skipToMainContent` class to be `10000px` instead of `-10000px`. This places the skip link on the left side of the sidebar content instead of the right.

I like the second option, and I think it's what I meant to when I implemented this originally!

**Comment by ashleemboyer** [comment]
One thing I forgot to mention yesterday in my review: the heading links aren't quite working when you try to navigate to them directly or when you try to go to them from the sidebar. Here's an example link: https://dev-portal-git-zsrefactor-sidebar-layout-hashicorp.vercel.app/waypoint#featured-reference-docs. I think this has something to do with the page content now going behind the sticky header?

File: src/layouts/docs/partials/sidebar-sidecar-layout/style.module.css
Comment (zchsh) [reply src/layouts/docs/partials/sidebar-sidecar-layout/style.module.css]: @ashleemboyer Sounds good, thanks for the added thoughts - I've moved `background-color` to `sidebar` 👍

Comment (ashleemboyer) [reply src/layouts/docs/partials/sidebar-sidecar-layout/style.module.css]: Awesome! I'll mark this as resolved 😊

**Comment by zchsh** [comment]
@ashleemboyer Great catch re: scroll offset. Would it be alright with you if we handle that concern in a separate PR? I think this is [an issue on our existing sites as well](https://www.vaultproject.io/docs/install#linux-package), just wary of having the scope of this PR expand further and get stuck. Let me know what you think! 🙇

**Comment by ashleemboyer** [comment]
> @ashleemboyer Great catch re: scroll offset. Would it be alright with you if we handle that concern in a separate PR? I think this is [an issue on our existing sites as well](https://www.vaultproject.io/docs/install#linux-package), just wary of having the scope of this PR expand further and get stuck. Let me know what you think! 🙇

I'm not seeing an offset when I go to that URL, and I think it depends if we should solve here or not. There isn't currently an offset issue so it's something about the refactor that introduced it. I'm not sure if we should merge the regression if we don't have a clear idea of what to try to fix it. What do you think?

**Comment by zchsh** [comment]
@ashleemboyer Pushed up an offset fix, I thought this was something we'd done on the docs sites, but when I visit https://www.vaultproject.io/docs/install#linux-package I'm seeing the `Linux package` heading hidden under the top sticky navs (should have been more clear about this - if you're not seeing the same let me know, maybe there's a fix we have in place that I'm somehow missing!).

Going to get those upstream changes in to fix the conflicts since #79 merged, and then I _think_ this should be ready for re-review - I believe one outstanding item is how we're structuring or naming `docs-layout` vs `sidebar-sidecar-layout`.

My intent with the current `sidebar-sidecar` naming, and separate partial, is that we may be able to re-use it for the `Install` view. That view seems to follow the same layout, but with different sidecar contents. Definitely open to refining the layout organization further in further PRs. Let me know what you think! 🙇

**Comment by ashleemboyer** [comment]
> @ashleemboyer Pushed up an offset fix, I thought this was something we'd done on the docs sites, but when I visit https://www.vaultproject.io/docs/install#linux-package I'm seeing the `Linux package` heading hidden under the top sticky navs (should have been more clear about this - if you're not seeing the same let me know, maybe there's a fix we have in place that I'm somehow missing!).

Weird! I'm not seeing that. Here's a screen recording of what I see. Maybe it's my screen size or that I'm using Chrome?

https://user-images.githubusercontent.com/43934258/151254323-abac4833-52a7-4a82-a9ed-02a4b07e2676.mov

> Going to get those upstream changes in to fix the conflicts since #79 merged, and then I _think_ this should be ready for re-review - I believe one outstanding item is how we're structuring or naming `docs-layout` vs `sidebar-sidecar-layout`.
>
> My intent with the current `sidebar-sidecar` naming, and separate partial, is that we may be able to re-use it for the `Install` view. That view seems to follow the same layout, but with different sidecar contents. Definitely open to refining the layout organization further in further PRs. Let me know what you think! 🙇

Thanks for elaborating! That makes sense. My plan for `Sidecar` with that new view (which I haven't brought up in any discussions yet) is to refactor `Sidecar` to accept children instead of only rendering one thing. Then I'm thinking to create a `TableOfContents` component (or something of that name) that handles rendering the `<nav>` element currently in the `Sidecar` and passing that as a child. With this refactor in mind, I think renaming `docs-layout` to `sidebar-sidecar-layout` is a good approach, then everything is abstracted at a nice level and everything (except `Sidebar` lol) is essentially just a container rendered in the correct place on the page by `sidebar-sidecar-layout`.

**Comment by zchsh** [comment]
@ashleemboyer Ah neat, yep, I'm seeing the same thing in Chrome and Firefox (heading visible, which is nice), just seems like Safari where I'm seeing the hidden heading.

Does the fix for heading offsets work for you? Eg does https://dev-portal-git-zsrefactor-sidebar-layout-hashicorp.vercel.app/waypoint#featured-reference-docs work? (Seems to work for me in Safari, Chrome, & Firefox 🤞. Also want to acknowledge it's not the best long-term fix; my hope is we can iterate on it as part of the general `.__permalink-h` refactor work 👍)

Re: `DocsLayout` renaming, do you mind if we handle that in your Install view PR? Definitely open to the rename etc, just having trouble figuring out how all those pieces fit together without the extra Install page in place (eg I assumed we'd pull out `sidebar-sidecar-layout` from `layouts/docs/partials`, and leave eg `layouts/docs/server` in place; I think those concerns will be easier to work through with more concrete reasons to move things around in front of us). Let me know if that's alright! 🙇‍♂️

**Comment by zchsh** [comment]
Not sure what's going on with the Bundle Analysis check here 🤔 Seems unrelated to this PR I think?

File: src/components/sidebar/components/sidebar-nav/style.module.css
Comment (ashleemboyer) [reply src/components/sidebar/components/sidebar-nav/style.module.css:6]: @zchsh looks like this still needs to be addressed! 🙏🏻

File: src/layouts/docs/docs-layout.module.css
Comment (ashleemboyer) [inline src/layouts/docs/docs-layout.module.css:72]: This is really neat! Nicely done ✨

**Review by ashleemboyer — CHANGES_REQUESTED** [review]
In addition to the things we've already chatted about via Slack, just seeing that one more thing needs addressed! We can defer the renaming of the layout, potentially even when I work on refactoring Sidecar.

File: src/layouts/docs/docs-layout.module.css
Comment (BRKalow) [reply src/layouts/docs/docs-layout.module.css:72]: for consideration: I think something similar to this is how we're handling it on the current docs sites: https://github.com/hashicorp/nomad/blob/main/website/pages/style.css#L33-L49

File: src/components/sidebar/components/sidebar-nav/style.module.css
Comment (zchsh) [reply src/components/sidebar/components/sidebar-nav/style.module.css]: Ah thanks for flag, lost track of this conversation! Just adjusted with the second option 💯 ✅

**Review by ashleemboyer — CHANGES_REQUESTED** [review]
Latest updates are looking great! It uncovered one more bug I wasn't noticing before. The `<nav>` in sidecar didn't used to have a parent container, and it has two classes on it that handle making the element go away on narrower views: `hide-on-mobile` and `hide-on-tablet`. Those classes need to be moved to the new containing `<div>` around that `<nav>` or we see something like this, an empty div that's `220px` wide:

image

**Comment by zchsh** [comment]
@ashleemboyer Thanks for catching the sidecar bug - should be fixed now! 🙇

**Review by ashleemboyer — APPROVED** [review]
Woo! 🥇

**Pull request merged**
